packages/miniflare/src/shared/dev-registry.worker.ts

```diff
@@ -0,0 +1,349 @@
+			});
+		} catch (e) {
+			log.error(e instanceof Error ? e : new Error(`${e}`));
+			clientSocket.end();
+		}
+	}
+
+	private getServiceAddress(service: string, entrypoint: string): ProxyAddress {
+		const target = this.registry?.[service];
+		const entrypointAddress = target?.entrypointAddresses[entrypoint];
+
+		if (entrypointAddress !== undefined) {
+			return {
+				httpStyle: "proxy",
+				protocol: target.protocol,
+				host: entrypointAddress.host,
+				port: entrypointAddress.port,
+			};
+		}
+
+		if (target && target.protocol !== "https" && entrypoint === "default") {
+			// Fallback to sending requests directly to the entry worker
+			return {
+				httpStyle: "host",
+				protocol: target.protocol,
+				host: target.host,
+				port: target.port,
+			};
+		}
+
+		return this.getFallbackServiceAddress(service, entrypoint);
+	}
+
+	private getFallbackServiceAddress(
+		service: string,
+		entrypoint: string
+	): ProxyAddress {
+		if (!this.runtimeEntryURL) {
+			throw new Error("No runtime entry URL set");
+		}
+
+		const url = new URL(this.runtimeEntryURL);
+		const port = this.fallbackServicePorts[service]?.[entrypoint];
+
+		if (!port) {
+			throw new Error(
+				`There is no socket opened for "${service}" with the "${entrypoint}" entrypoint`
+			);
+		}
+
+		return {
+			httpStyle: "proxy",
+			protocol: getProtocol(url),
+			host: url.hostname,
+			port,
+		};
+	}
+
+	private getDurableObjectAddress(
+		scriptName: string,
+		className: string
+	): ProxyAddress | null {
+		const target = this.registry?.[scriptName];
+
+		if (
+			target?.durableObjects.some(
+				(durableObject) => durableObject.className === className
+			)
+		) {
+			return {
+				...target,
+				path: `/${INBOUND_DO_PROXY_SERVICE_PATH}`,
+			};
+		}
+
+		return null;
+	}
+
+	private handleProxy(
+		req: http.IncomingMessage,
+		res: http.ServerResponse,
+		target: ProxyAddress
+	) {
+		const headers = { ...req.headers };
+		let path = target.path;
+
+		if (!path) {
+			switch (target.httpStyle) {
+				case "host": {
+					const url = new URL(req.url ?? `http://${req.headers.host}`);
+					// If the target is a host, use the path from the request URL
+					path = url.pathname + url.search + url.hash;
+					headers.host = url.host;
+					break;
+				}
+				case "proxy": {
+					// If the target is a proxy, use the full request URL
+					path = req.url;
+					break;
+				}
+			}
+		}
+
+		const options: http.RequestOptions = {
+			host: target.host,
+			port: target.port,
+			method: req.method,
+			path,
+			headers,
+		};
+
+		const upstream = http.request(options, (upRes) => {
+			// Relay status and headers back to the original client
+			res.writeHead(upRes.statusCode ?? 500, upRes.headers);
+			// Pipe the response body
+			upRes.pipe(res);
+		});
+
+		// Pipe the client request body to the upstream
+		req.pipe(upstream);
+
+		upstream.on("error", () => {
+			if (!res.headersSent) res.writeHead(502);
+			res.end("Bad Gateway");
+		});
+	}
```

Comment:
This was moved from the [miniflare class](https://github.com/cloudflare/workers-sdk/pull/10142/files#r2252264747).